airflow/models.py

```diff
@@ -2941,8 +3047,56 @@ def __init__(self, task):
         Trigger.__init__(self, task)
 
     def evaluate(self, execution_date, session):
-        return True, "DummyTrigger always evaluates to True"
+        reason = Reason(self, execution_date, True,
+                        comment="DummyTrigger always evaluates to True")
+        return True, reason
 
     def __repr__(self):
         return "<DummyTrigger: {dag}.{task}>" \
                "".format(dag=self.task.dag_id, task=self.task.task_id)
+
+
+class Reason(object):
+
+    def __init__(self,
+                 trigger,
+                 execution_date,
+                 result,
+                 states=None,
+                 comment=""):
+        self.trigger = trigger
+        self.execution_date = execution_date
+        self.result = result
+        self.states = states if states else {}
+        self.comment = comment
+
+    def is_failed(self):
+        # if the result is False
+        bad_result = not self.result
+
+        # if all the dependencies are done (= have already run)
+        all_dependencies_done = all([state in State.done()
+                                     for state in self.states.values()])
+        # if there's a failed state
+        failed_state = \
+            State.FAILED in self.states.values() or \
+            State.UPSTREAM_FAILED in self.states.values()
+
+        return bad_result and all_dependencies_done and failed_state
+
+    def is_skipped(self):
+        # if the result is False and
+        bad_result = not self.result
+
+        # if all the dependencies are done (= have already run) and
+        all_dependencies_done = all([state in State.done()
+                                     for state in self.states.values()])
+        # if the reason is not is_failed()
```

Comment:
we should probably factor out the logic of `all_dependencies_done` into a re-usable and unit testable method, I think it's present at 3 places at the moment